File: app/shared/src/commonTest/kotlin/com/tneff/cyppieagents/multihub/ActiveHubConnectionTest.kt
package com.tneff.cyppieagents.multihub

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * CYP-865 — the switch-first / ONE-ACTIVE-HUB lifecycle: `switchTo` tears down the OLD connection BEFORE opening the
 * new one (exactly one live at a time; no background connection to an inactive hub), switch-to-active is a no-op (no
 * teardown + re-dial), and `close` tears down. The connector is an injected fake (arming DARK — no real dial).
 */
class ActiveHubConnectionTest {

    /** A fake handle whose `close()` and the connector's `open()` append to a shared log (order-sensitive proof). */
    private class FakeHandle(private val id: String, private val log: MutableList<String>) : HubConnectionHandle<String> {
        override val machine: String = id
        override fun close() { log.add("close:$id") }
    }

    private fun loggingConnector(log: MutableList<String>) = HubConnector<String> { hubId ->
        log.add("open:$hubId")
        FakeHandle(hubId, log)
    }

    @Test
    fun active_isNull_beforeFirstConnect() {
        val ahc = createActiveHubConnection(loggingConnector(mutableListOf()))
        assertNull(ahc.active())
    }

    @Test
    fun switchTo_opensAndBecomesActive() {
        val log = mutableListOf<String>()
        val ahc = createActiveHubConnection(loggingConnector(log))
        ahc.switchTo("a")
        assertEquals(listOf("open:a"), log)
        assertEquals("a", ahc.active()?.hubId)
        assertEquals("a", ahc.active()?.machine)
    }

    @Test
    fun switchTo_tearsDownOld_beforeOpeningNew() {
        val log = mutableListOf<String>()
        val ahc = createActiveHubConnection(loggingConnector(log))
        ahc.switchTo("a")
        ahc.switchTo("b")
        // The OLD hub's connection is closed BEFORE the new one opens — exactly one live connection at a time.
        assertEquals(listOf("open:a", "close:a", "open:b"), log)
        assertEquals("b", ahc.active()?.hubId)
    }

    @Test
    fun switchTo_alreadyActive_isNoOp_noTeardownNoReDial() {
        val log = mutableListOf<String>()
        val ahc = createActiveHubConnection(loggingConnector(log))
        ahc.switchTo("a")
        log.clear()
        ahc.switchTo("a")
        assertTrue(log.isEmpty(), "switch-to-active must NOT tear down + re-dial the live connection")
        assertEquals("a", ahc.active()?.hubId)
    }

    @Test
    fun close_tearsDown_andClearsActive() {
        val log = mutableListOf<String>()
        val ahc = createActiveHubConnection(loggingConnector(log))
        ahc.switchTo("a")
        log.clear()
        ahc.close()
        assertEquals(listOf("close:a"), log)
        assertNull(ahc.active())
    }
}

File: web-ts/src/state/activeHubConnection.test.ts
import { describe, it, expect } from 'vitest'
import { createActiveHubConnection } from './activeHubConnection'
import { createRemoteConnMachine, failureDisposition } from './remoteConnState'
import { FakeRemoteConnector } from './testing/fakeRemoteConnector'

/** A recording connector: tracks which hubs were opened and which were torn down (close()). */
const recordingConnector = (opens: string[], closes: string[]) => (hubId: string) => {
  opens.push(hubId)
  return { machine: createRemoteConnMachine(), close: () => closes.push(hubId) }
}

describe('CYP-853 (Multi-Hub M3) — activeHubConnection: ONE-ACTIVE switch-first lifecycle', () => {
  it('switchTo opens a live connection for the target hub', () => {
    const opens: string[] = []
    const ahc = createActiveHubConnection(recordingConnector(opens, []))
    ahc.switchTo('local')
    expect(opens).toEqual(['local'])
    expect(ahc.active()?.hubId).toBe('local')
  })

  it('★ switching TEARS DOWN the old before opening the new — exactly ONE live at a time (no background conn survives)', () => {
    // MUT: open the new without closing the old → the inactive hub keeps a background connection (breaks switch-first,
    // and the immediate-STALE precondition) → closes stays [] → reds.
    const opens: string[] = []
    const closes: string[] = []
    const ahc = createActiveHubConnection(recordingConnector(opens, closes))
    ahc.switchTo('local')
    ahc.switchTo('remote')
    expect(closes).toEqual(['local']) // the OLD hub's connection was torn down
    expect(opens).toEqual(['local', 'remote'])
    expect(opens.length - closes.length).toBe(1) // exactly one live connection
    expect(ahc.active()?.hubId).toBe('remote')
  })

  it('★ switch-to-active is a no-op: the live connection is NOT torn down and re-dialed', () => {
    // MUT: drop the same-hub guard → switching to the current hub closes + re-opens it (a needless reconnect) → reds.
    const opens: string[] = []
    const closes: string[] = []
    const ahc = createActiveHubConnection(recordingConnector(opens, closes))
    ahc.switchTo('local')
    ahc.switchTo('local')
    expect(opens).toEqual(['local'])
    expect(closes).toEqual([])
  })

  it('close() tears down the active connection and clears it', () => {
    const closes: string[] = []
    const ahc = createActiveHubConnection(recordingConnector([], closes))
    ahc.switchTo('local')
    ahc.close()
    expect(closes).toEqual(['local'])
    expect(ahc.active()).toBeNull()
  })

  it('★ MC-6: the active connection surfaces an ACTIONABLE failure (remote-not-configured) through a real machine', () => {
    // The conn-model drives a REAL createRemoteConnMachine; a remote-not-configured verdict lands failed(actionable),
    // not terminal/retryable. MUT: a connector that swallowed the verdict / a wrong disposition mapping → reds.
    const fake = new FakeRemoteConnector()
    const ahc = createActiveHubConnection(() => {
      const machine = createRemoteConnMachine()
      const unsub = fake.subscribe(machine.send)
      return { machine, close: unsub }
    })
    ahc.switchTo('remote')
    fake.dial()
    fake.openHandshake()
    fake.completeHandshake()
    fake.evaluateTrustFromIssuer('REMOTE_NOT_CONFIGURED')
    const st = ahc.active()!.machine.getState()
    expect(st.phase).toBe('failed')
    if (st.phase === 'failed') {
      expect(st.cause).toBe('remote-not-configured')
      expect(failureDisposition(st.cause)).toBe('actionable') // not terminal, not retryable
    }
  })
})
